expose error handling for state changes

Karsten1987 opened the pull request:
two changes in this PR:

1.) remove `fprintf`s and replace them with `RCUTILS_LOG..`
2.) every state change has now an additional parameter to see whether the action went to an erroneous state while transitioning. 

CI:
linux: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2917)](http://ci.ros2.org/job/ci_linux/2917/)
osx: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=2361)](http://ci.ros2.org/job/ci_osx/2361/)
windows: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3041)](http://ci.ros2.org/job/ci_windows/3041/)

wjwwood commented on rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp:
```diff
@@ -179,7 +181,10 @@ class LifecycleNode::LifecycleNodeInterfaceImpl
-    resp->success = change_state(req->transition.id);
+    rcl_lifecycle_ret_t error;
+    auto ret = change_state(req->transition.id, error);
+    (void) error;
```
Maybe you should call this `callback_return_code` or something. `error` is not very descriptive.

wjwwood commented on rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp:
```diff
@@ -273,63 +277,61 @@ class LifecycleNode::LifecycleNodeInterfaceImpl
     // error handling ?!
     // TODO(karsten1987): iterate over possible ret value
     if (cb_success == RCL_LIFECYCLE_RET_ERROR) {
+      RCUTILS_LOG_WARN("Error occured. Executing error handling.")
```
nitpick, I don't think this should be two sentences, instead maybe "Error occurred while doing error handling". (also occurred is misspelled)

dirk-thomas commented on rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp:
```diff
@@ -373,29 +373,57 @@ class LifecycleNode : public node_interfaces::LifecycleNodeInterface,
 
   RCLCPP_LIFECYCLE_PUBLIC
   const State &
+  trigger_transition(const Transition & transition, rcl_ret_t & error);
```
Why is this function getting a `rcl_ret_t` argument while the other functions get a `rcl_lifecycle_ret_t`?

dirk-thomas commented on rclcpp_lifecycle/src/lifecycle_node.cpp:
```diff
@@ -306,46 +306,93 @@ LifecycleNode::trigger_transition(const Transition & transition)
 }
 
 const State &
+LifecycleNode::trigger_transition(uint8_t transition_id, rcl_ret_t & error)
```
This signature doesn't match the header (`rcl_ret_t` vs. `rcl_lifecycle_ret_t`)?

dirk-thomas commented on rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp:
```diff
@@ -179,7 +181,10 @@ class LifecycleNode::LifecycleNodeInterfaceImpl
+    rcl_lifecycle_ret_t error;
+    auto ret = change_state(req->transition.id, error);
+    (void) error;
+    resp->success = (ret == RCL_RET_OK) ? true : false;
```
The expression in the parenthesis is already the boolean you want. No need to the ternary operator here.

dirk-thomas commented on rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp:
```diff
@@ -273,63 +277,61 @@ class LifecycleNode::LifecycleNodeInterfaceImpl
         if (rcl_lifecycle_trigger_transition(&state_machine_, error_resolved, true) != RCL_RET_OK) {
-          fprintf(stderr, "Failed to call cleanup on error state\n");
-          return false;
+          RCUTILS_LOG_ERROR("Failed to call cleanup on error state");
```
No semicolon after logging macros.

Same below.

Karsten1987 commented:
I hopefully addressed all comments. If nobody has further comments until tomorrow afternoon, i'll go ahead and merge.

ran a new batch of ci:

linux: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2918)](http://ci.ros2.org/job/ci_linux/2918/)
osx: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=2362)](http://ci.ros2.org/job/ci_osx/2362/)
win: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3042)](http://ci.ros2.org/job/ci_windows/3042/)
